internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,424 @@
+	ReplicaDiscoveryEnabled    bool
+	SpecifiedClient            *azappconfig.Client
+	DynamicClients             []*ConfigurationClientWrapper
+	ValidDomain                string
+	Endpoint                   string
+	ClientsFallbackStatus      map[string]*ConfigurationClientWrapper
+	Credential                 azcore.TokenCredential
+	ConnectionString           string
+	secret                     string
+	id                         string
+	lastFallbackClientsAttempt metav1.Time
+	lastFallbackRefresh        metav1.Time
+}
+
+type ConfigurationClientWrapper struct {
+	Endpoint       string
+	Client         *azappconfig.Client
+	BackOffEndTime metav1.Time
+	FailedAttempts int
+}
+
+type ConfigClientManager interface {
+	NewConfigurationClientManager(ctx context.Context, provider acpv1.AzureAppConfigurationProvider) (*ConfigurationClientManager, error)
+	GetClients() []*ConfigurationClientWrapper
+	DiscoverFallbackClients(host string) error
+}
+
+const (
+	TCP                                 string        = "tcp"
+	Origin                              string        = "origin"
+	Alt                                 string        = "alt"
+	EndpointSection                     string        = "Endpoint"
+	SecretSection                       string        = "Secret"
+	IdSection                           string        = "Id"
+	AzConfigDomainLabel                 string        = "azconfig"
+	AppConfigDomainLabel                string        = "appconfig"
+	FallbackClientRefreshExpireInterval time.Duration = time.Hour
+	MinimalClientRefreshInterval        time.Duration = time.Second * 30
+	MaxBackoffDuration                  time.Duration = time.Minute * 10
+	MinBackoffDuration                  time.Duration = time.Second * 30
+	JitterRatio                         float64       = 0.25
+	SafeShiftLimit                      int           = 63
+)
+
+var (
+	clientOptionWithModuleInfo *azappconfig.ClientOptions = &azappconfig.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Telemetry: policy.TelemetryOptions{
+				ApplicationID: fmt.Sprintf("%s/%s", properties.ModuleName, properties.ModuleVersion),
+			},
+		},
+	}
+)
+
+func NewConfigurationClientManager(ctx context.Context, provider acpv1.AzureAppConfigurationProvider) (*ConfigurationClientManager, error) {
+	manager := &ConfigurationClientManager{
+		ReplicaDiscoveryEnabled: provider.Spec.ReplicaDiscovery,
+		ValidDomain:             getValidDomain(*provider.Spec.Endpoint),
+		Endpoint:                *provider.Spec.Endpoint,
+		ClientsFallbackStatus:   make(map[string]*ConfigurationClientWrapper),
+	}
+
+	var err error
+	if provider.Spec.ConnectionStringReference != nil {
+		manager.ConnectionString, err = getConnectionStringParameter(ctx, types.NamespacedName{Namespace: provider.Namespace, Name: *provider.Spec.ConnectionStringReference})
+		if err != nil {
+			return nil, err
+		}
+		manager.secret, err = parseConnectionString(manager.ConnectionString, SecretSection)
```

Comment:
why we keep both secret/id and connectionString? looks like one of them is enough.